djangosaml2idp/models.py

```diff
@@ -51,26 +51,56 @@ def refresh_metadata(self, force_refresh: bool = False) -> bool:
-
-            # TODO: commented this out to pass test
-            # self.metadata_expiration_dt = extract_validuntil_from_metadata(self.local_metadata)
-            return True
-        return False
+        # Refresh in these cases:
+        should_refresh = False
+        # - Data was not fetched ever before, so local_metadata is empty
+        if not self.local_metadata:
+            should_refresh = True
+        # - The expiration timestamp is not set or expired
+        if not self.metadata_expiration_dt or now() > self.metadata_expiration_dt:
+            should_refresh = True
+        # - Force refresh requested by caller
+        if force_refresh:
+            should_refresh = True
+
+        if not should_refresh:
+            return False
+
+        # Try to do the refresh if a remote is set
+        if self.remote_metadata_url:
+            try:
+                self.local_metadata = validate_metadata(fetch_metadata(self.remote_metadata_url))
+                self.metadata_expiration_dt = extract_validuntil_from_metadata(self.local_metadata).replace(tzinfo=None)
+                # Return True if it is now valid, False (+ log an error) otherwise
+                if now() > self.metadata_expiration_dt:
+                    logger.error(f'Remote metadata for SP {self.entity_id} was refreshed, but contains an expired validity datetime.')
+                    return False
+                return True
+            except Exception as e:
+                logger.error(f'Metadata for SP {self.entity_id} could not be pulled from remote url {self.remote_metadata_url}.', extra={'exception': str(e)})
+                return False
+
+        # No remote metadata url to fetch new data from. Log an error anyways in case
+        # - the current local data is not present
+        # - if it does not have a valid expiry timestamp
+        # - if it is expired
+        if not self.local_metadata:
+            logger.error(f'Local metadata for SP {self.entity_id} is not present, and no remote metadata found to refresh.')
+            return False
+        if (not self.metadata_expiration_dt) or (now() > self.metadata_expiration_dt):
+            try:
+                # Try to extract a valid expiration datetime from the local metadata succesfully
+                self.metadata_expiration_dt = extract_validuntil_from_metadata(self.local_metadata).replace(tzinfo=None)
+                # Return True if it is now valid, False (+ log an error) otherwise
+                if now() > self.metadata_expiration_dt:
+                    logger.error(f'Local metadata for SP {self.entity_id} contains an expired validity datetime or none at all, no remote metadata found to refresh.')
+                    return False
+                return True
+            except Exception as e:
+                logger.error(f'Metadata expiration dt for SP {self.entity_id} could not be extracted from local metadata.', extra={'exception': str(e)})
+                return False
```

Comment:
I would refactor both of these `try/except` blocks into their own private methods. Would make it easier for testing as well as readability.

```python
if self.remote_metadata_url:
    return self._refresh_from_remote()

if (not self.metadata_expiration_dt) or (now() > self.metadata_expiration_dt):
    return self._refresh_from_local()
```